packages/subgraph/src/utils/tokens/erc20.ts

```diff
@@ -0,0 +1,217 @@
+  let totalSupply = erc20.try_totalSupply();
+
+  if (totalSupply.reverted) {
+    return null;
+  }
+  contract.name = try_name.reverted ? '' : try_name.value;
+  contract.symbol = try_symbol.reverted ? '' : try_symbol.value;
+
+  contract.save();
+
+  return contract;
+}
+
+export function updateERC20Balance(
+  token: Address,
+  dao: Address,
+  amount: BigInt,
+  timestamp: BigInt,
+  type: TransferType
+): void {
+  let daoId = dao.toHexString();
+  let balanceId = daoId.concat('_').concat(token.toHexString());
+
+  let erc20Balance = ERC20Balance.load(balanceId);
+
+  if (!erc20Balance) {
+    erc20Balance = new ERC20Balance(balanceId);
+    erc20Balance.dao = daoId;
+    erc20Balance.token = token.toHexString();
+    erc20Balance.balance = BigInt.zero();
+  }
+
+  let erc20 = ERC20.bind(token);
+  let balance = erc20.try_balanceOf(dao);
+  if (!balance.reverted) {
+    erc20Balance.balance = balance.value;
+  }
+
+  erc20Balance.lastUpdated = timestamp;
+  erc20Balance.save();
+}
+
+export function handleERC20Action(
+  token: Address,
+  dao: Address,
+  proposalId: string,
+  data: Bytes,
+  actionIndex: number,
+  event: ethereum.Event
+): void {
+  let contract = fetchERC20(token);
+  if (!contract) {
+    return;
+  }
+
+  let decodeABI = '';
+
+  let functionSelector = data.toHexString().substring(0, 10);
+  let calldata = data.toHexString().slice(10);
+
+  if (functionSelector == ERC20_transfer) {
+    decodeABI = '(address,uint256)';
+  }
+
+  if (functionSelector == ERC20_transferFrom) {
+    decodeABI = '(address,address,uint256)';
+  }
+
+  let decoded = ethereum.decode(decodeABI, Bytes.fromHexString(calldata));
+
+  if (!decoded) {
+    return;
+  }
+
+  let tuple = decoded.toTuple();
+
+  let from = new Address(0);
+  let to = new Address(0);
+  let amount = BigInt.zero();
+
+  if (functionSelector == ERC20_transfer) {
+    from = dao;
+    to = tuple[0].toAddress();
+    amount = tuple[1].toBigInt();
+  }
+
+  if (functionSelector == ERC20_transferFrom) {
+    from = tuple[0].toAddress();
+    to = tuple[1].toAddress();
+    amount = tuple[2].toBigInt();
+  }
+
+  // Ambiguity ! No need to store transfer such as this.s
+  if (from == to) {
+    return;
+  }
+
+  let daoId = dao.toHexString();
+
+  let id = getTransferId(
+    event.transaction.hash,
+    event.transactionLogIndex,
+    actionIndex
+  );
+
+  let erc20Transfer = new ERC20Transfer(id);
+
+  erc20Transfer.from = from;
+  erc20Transfer.to = to;
+  erc20Transfer.dao = daoId;
+  erc20Transfer.amount = amount;
+  erc20Transfer.txHash = event.transaction.hash;
+  erc20Transfer.createdAt = event.block.timestamp;
+  erc20Transfer.token = contract.id;
+  erc20Transfer.proposal = proposalId;
+
+  // If from/to both aren't equal to dao, it means
+  // dao must have been approved for the `amount`
+  // and played the role of transfering between 2 parties.
+  if (from != dao && to != dao) {
+    erc20Transfer.type = 'None'; // No idea
```

Comment:
Wouldn't this mean the DAO has sent money on behalf of someone else that approved the DAO as a spender?

Not sure if we should still store it as the DAO acted / did something